File: Test/Test algorithme partage des eaux/LPE.py
from ListPixel import ListPixel
from Pixel import Pixel
from MorphologiquesOperations import MorphologiquesOperations
from Image import Image
import matplotlib.image as mpimg

class LPE():
	def __init__(self):
		self.INIT = -1
		self.MASK = -2
		self.WSHED = 0
		self.MARK = -1

		self.histogramme = [0]*256
		self.image = Image()
		self.sortedImage = Image()
		self.sortedPixels = ListPixel()

		self.pFictif = Pixel(-1, -1, -1)



	def init(self, image):
		self.image.init1(image)
		self.histo()
		print("histo")
		self.sortImage()
		print("sort")

	def histo(self):
		for level in range(256):
			self.histogramme[level] = self.image.getNumberPixelsAtLevel(level)

	def sortImage(self):
		self.sortedPixels.addPixels(self.image.getImage())


	def isNeighbour(self, h, w, row, column, varRow, varCol):
		if((row+varRow>=0) and (row+varRow < h) and (column+varCol>=0) and (column+varCol < w)):
			return ((varRow == 0 and varCol != 0) or (varRow != 0 and varCol == 0))
		return False

	def runLPE(self):
		lab = Image()
		lab.init3(self.image.getHeight(), self.image.getWidth(), -1)
		dist = Image()
		dist.init2(self.image.getHeight(), self.image.getWidth())

		curlab = 0

		pile = []

		for h in range(256):
			print(h)

			pixels = self.sortedPixels.getPixelsByLevel(h)
			for curPix in pixels:
				lab.write1(curPix.getX(), curPix.getY(), -2)
				row = curPix.getX()
				column = curPix.getY()

				done = True
				for varRow in range(-1, 2, 1):
					if(done):
						for varCol in range(-1, 2, 1):
							if (done):
								if(self.isNeighbour(lab.getHeight(), lab.getWidth(), row, column, varRow, varCol)):
									if(lab.read1(varRow+row, varCol+column) > 0 or lab.read1(varRow+row, varCol+column) == self.WSHED):
										done = False
										dist.write1(curPix.getX(), curPix.getY(), 1)
										pile.append(curPix)
			curDist = 1
			pile.append(self.pFictif)

			while 1:
				curPix = pile.pop(0)
				if(curPix.getX() == -1 and curPix.getY() == -1):
					if(len(pile) == 0):
						break
					pile.append(self.pFictif)
					curDist += 1
					curPix = pile.pop(0)

				row = curPix.getX()
				column = curPix.getY()

				for varRow in range(-1, 2, 1):
					for varCol in range(-1, 2, 1):
						if(self.isNeighbour(lab.getHeight(), lab.getWidth(), row, column, varRow, varCol)):
							neighbourPos = lab.getAbsolutePosition(row+varRow, column+varCol)
							if((dist.read2(neighbourPos)< curDist) and (lab.read2(neighbourPos) > 0 or lab.read2(neighbourPos) == self.WSHED)):
								if(lab.read2(neighbourPos)>0):
									if(lab.read1(curPix.getX(), curPix.getY()) == -2 or lab.read1(curPix.getX(), curPix.getY()) == self.WSHED):
										lab.write1(curPix.getX(), curPix.getY(), lab.read2(neighbourPos))
									elif (lab.read1(curPix.getX(), curPix.getY()) != lab.read2(neighbourPos)):
										lab.write1(curPix.getX(), curPix.getY(), self.WSHED)
										#lab.write2(neighbourPos, curlab)
								elif(lab.read1(curPix.getX(), curPix.getY()) == -2):
									lab.write1(curPix.getX(), curPix.getY(), self.WSHED)
									#lab.write2(neighbourPos, curlab)
							elif(lab.read2(neighbourPos) == -2 and dist.read2(neighbourPos) == 0):
								dist.write2(neighbourPos, curDist+1)
								pile.append(Pixel(self.image.getRowFromAbsolutePosition(neighbourPos),
								 self.image.getColumnFromAbsolutePosition(neighbourPos), 0))

			for curPix in pixels:
				dist.write1(curPix.getX(), curPix.getY(), 0)
				if(lab.read1(curPix.getX(), curPix.getY()) == -2):
					curlab += 1
					pile.append(curPix)
					lab.write1(curPix.getX(), curPix.getY(), curlab)
					while(len(pile) != 0):
						curPix = pile.pop(0)
						row2 = curPix.getX()
						column2 = curPix.getY()
						for varRow in range(-1, 2, 1):
							for varCol in range(-1, 2, 1):
								if(self.isNeighbour(lab.getHeight(), lab.getWidth(), row2, column2, varRow, varCol)):
									neighbourPos = lab.getAbsolutePosition(row2+varRow, column2+varCol)
									if(lab.read2(neighbourPos) == -2):
										pile.append(Pixel(self.image.getRowFromAbsolutePosition(neighbourPos),
											self.image.getColumnFromAbsolutePosition(neighbourPos), 0))
										lab.write2(neighbourPos, curlab)

		print("Segmentation")
		if(True):
			for curPix in range(lab.getSurface()):
				if(lab.read2(curPix)>0):
					row = lab.getRowFromAbsolutePosition(curPix)
					column = lab.getColumnFromAbsolutePosition(curPix)

					done = True
					for varRow in range(-1, 2, 1):
						if(done):
							for varCol in range(-1, 2, 1):
								if (done):
									if(self.isNeighbour(lab.getHeight(), lab.getWidth(), row, column, varRow, varCol)):
										neighbourPos = lab.getAbsolutePosition(row+varRow, column+varCol)
										if(lab.read2(curPix) > lab.read2(neighbourPos) and lab.read2(neighbourPos)>0):
											lab.write2(curPix, 0)
											done = False
		return lab

File: Test/Test algorithme partage des eaux/classe/lepe/LPE.class_source_from_jdcore/LPE.java
package lpe;

import java.util.Vector;












public class LPE
{
  public static final int INIT = -1;
  public static final int MASK = -2;
  public static final int WSHED = 0;
  private final int MARK = -1;
  


  private GrayImage in;
  


  public int numberLabel;
  


  public int[] histogramme;
  


  public GrayImage sortedImage;
  


  private int[] sortedPixels;
  


  public LPE(GrayImage gi)
  {
    in = gi;
    histo();
    sortImage();
  }
  



  private void histo()
  {
    int[] hist = new int['Ā'];
    
    for (int k = 0; k < 256; k++) {
      hist[k] = 0;
    }
    for (int i = 0; i < in.getHeight(); i++) {
      for (int j = 0; j < in.getWidth(); j++)
        hist[in.read(i, j)] += 1;
    }
    histogramme = hist;
  }
  



  private void sortImage()
  {
    GrayImage sgi = new GrayImage(in.getHeight(), in.getWidth());
    int px = 0;
    
    for (int k = 0; k < 256; k++) {
      for (int j = 0; j < histogramme[k]; j++) {
        sgi.write(px, k);
        px++;
      }
    }
    sortedImage = sgi;
    

    sortedPixels = new int[in.getSurface()];
    



    int[] index = new int[in.getSurface()];
    int s = 0;
    for (int i = 0; i < 256; i++) {
      index[i] = s;
      s += histogramme[i];
    }
    




    for (int pix = 0; pix < in.getSurface(); pix++) {
      int val = in.read(pix);
      int position = index[val];
      
      sortedPixels[position] = pix;
      index[val] += 1;
    }
  }
  



















  private boolean isNeighbour(int height, int width, int row, int column, int varRow, int varCol, boolean fourOrEight)
  {
    if ((row + varRow >= 0) && 
      (row + varRow < height) && 
      (column + varCol >= 0) && 
      (column + varCol < width)) {
      if (fourOrEight) {
        return 
          ((varRow == 0) && (varCol != 0)) || (
          (varRow != 0) && (varCol == 0));
      }
      return (varRow != 0) || (varCol != 0);
    }
    
    return false;
  }
  








  public GrayImage runLPE(boolean fourOrEight, boolean separation)
  {
    GrayImage lab = new GrayImage(in.getHeight(), in.getWidth());
    
    GrayImage dist = new GrayImage(in.getHeight(), in.getWidth());
    

    int curlab = 0;
    
    Fifo pile = new Fifo();
    

    for (int p = 0; p < in.getSurface(); p++) {
      lab.write(p, -1);
      dist.write(p, 0);
    }
    


    int curHistPos = 0;
    
    for (int h = 0; h < 256; h++) {
      for (int index = curHistPos; index < curHistPos + histogramme[h]; index++) {
        int curPix = sortedPixels[index];
        
        lab.write(curPix, -2);
        
        int row = lab.getRowFromAbsolutePosition(curPix);
        int column = lab.getColumnFromAbsolutePosition(curPix);
        
        boolean done = false;
        for (int varRow = -1; (varRow <= 1) && (!done); varRow++) {
          for (int varCol = -1; 
              (varCol <= 1) && (!done); 
              varCol++)
          {

            if (isNeighbour(lab.getHeight(), lab.getWidth(), row, column, varRow, varCol, fourOrEight)) {
              if ((lab.read(varRow + row, varCol + column) > 0) || 
                (lab.read(varRow + row, varCol + column) == 0))
              {
                dist.write(curPix, 1);
                pile.push(curPix);
                done = true;
              }
            }
          }
        }
      }
      int curDist = 1;
      
      pile.push(-1);
      
      for (;;)
      {
        int curPix = pile.pop();
        if (curPix == -1) {
          if (pile.isEmpty()) {
            break;
          }
          pile.push(-1);
          curDist++;
          curPix = pile.pop();
        }
        

        int row = lab.getRowFromAbsolutePosition(curPix);
        int column = lab.getColumnFromAbsolutePosition(curPix);
        
        for (int varRow = -1; varRow <= 1; varRow++) {
          for (int varCol = -1; varCol <= 1; varCol++) {
            if (isNeighbour(lab.getHeight(), lab.getWidth(), row, column, varRow, varCol, fourOrEight)) {
              int neighbourPos = 
                lab.getAbsolutePosition(
                row + varRow, 
                column + varCol);
              if ((dist.read(neighbourPos) < curDist) && (
                (lab.read(neighbourPos) > 0) || 
                (lab.read(neighbourPos) == 0))) {
                if (lab.read(neighbourPos) > 0) {
                  if ((lab.read(curPix) == -2) || 
                    (lab.read(curPix) == 0)) {
                    lab.write(
                      curPix, 
                      lab.read(neighbourPos));
                  }
                  else if (lab.read(curPix) != 
                    lab.read(neighbourPos)) {
                    lab.write(curPix, 0);
                  }
                }
                else if (lab.read(curPix) == -2) {
                  lab.write(curPix, 0);
                }
              }
              else if ((lab.read(neighbourPos) == -2) && 
                (dist.read(neighbourPos) == 0)) {
                dist.write(neighbourPos, curDist + 1);
                pile.push(neighbourPos);
              }
            }
          }
        }
      }
      
      for (int index = curHistPos; 
          index < curHistPos + histogramme[h]; 
          index++) {
        int curPix = sortedPixels[index];
        dist.write(curPix, 0);
        if (lab.read(curPix) == -2) {
          curlab++;
          pile.push(curPix);
          lab.write(curPix, curlab);
          int varRow; for (; !pile.isEmpty(); 
              




              varRow <= 1)
          {
            curPix = pile.pop();
            
            int row = lab.getRowFromAbsolutePosition(curPix);
            int column = lab.getColumnFromAbsolutePosition(curPix);
            
            varRow = -1; continue;
            for (int varCol = -1; varCol <= 1; varCol++) {
              if (isNeighbour(lab.getHeight(), 
                lab.getWidth(), 
                row, 
                column, 
                varRow, 
                varCol, 
                fourOrEight)) {
                int neighbourPos = 
                  lab.getAbsolutePosition(
                  row + varRow, 
                  column + varCol);
                if (lab.read(neighbourPos) == -2) {
                  pile.push(neighbourPos);
                  lab.write(neighbourPos, curlab);
                }
              }
            }
            varRow++;
          }
        }
      }
      


















      curHistPos += histogramme[h];
    }
    
    numberLabel = curlab;
    

    if (separation) {
      for (int curPix = 0; curPix < lab.getSurface(); curPix++) {
        if (lab.read(curPix) > 0)
        {
          int row = lab.getRowFromAbsolutePosition(curPix);
          int column = lab.getColumnFromAbsolutePosition(curPix);
          
          boolean done = false;
          for (int varRow = -1; 
              (varRow <= 1) && (!done); 
              varRow++) {
            for (int varCol = -1; 
                (varCol <= 1) && (!done); 
                varCol++) {
              if (isNeighbour(lab.getHeight(), 
                lab.getWidth(), 
                row, 
                column, 
                varRow, 
                varCol, 
                fourOrEight)) {
                int neighbourPos = 
                  lab.getAbsolutePosition(
                  row + varRow, 
                  column + varCol);
                if ((lab.read(curPix) > lab.read(neighbourPos)) && 
                  (lab.read(neighbourPos) > 0)) {
                  lab.write(curPix, 0);
                  done = true;
                }
              }
            }
          }
        }
      }
    }
    return lab;
  }
}
